File: aloelite/resolve.py
# ./aloelite/resolve.py
# License: Apache-2.0 (disclaimer at bottom of file)
"""
Path resolution — the first thing every implementation writes, and the thing
the whole flat layer is built on.

resolve() walks a whole path in ONE query via the `resolution.resolve_path`
recursive CTE, starting at the mount's mount point. resolve_parent() is the same
walk stopped one segment short, returning (container_id, final_name) for the
create / move / rename operations that need "the parent container, keep the
final name".

This used to fold `resolution.resolve_segment` per segment: a ten-deep path cost
ten round trips. Free on a local file, but path resolution is woven through every
path-addressed operation, so over a network connection to a remote backend the
per-segment fold is the difference between usable and unusable. The CTE resolves
any depth in one round trip and speeds up local operation too.

Path semantics are decided HERE, once, so all four implementations can mirror
exactly one set of rules:
  * paths are mount-relative; '' and '/' both denote the mount point itself
  * leading/trailing slashes are ignored; empty segments (//) collapse
  * resolution sees only VISIBLE nodes (NODE-5) — the greatest-uuid7 sibling
    wins — so hidden same-name siblings are unreachable by path. That is the
    contract; *_by_id variants exist for those.
  * a miss at any segment raises NotFound
  * a non-final segment that resolves to an entry (not a container) raises
    NotAContainer — you cannot descend through a file
  * '.' and '..' are ORDINARY NAMES, not navigation. They are never interpreted,
    so a path can never climb above the mount point it started at — which is
    what confines a subtree mount to its subtree. See
    tests/test_resolve_containment.py; that containment is a security boundary
    and must survive any future POSIX work.
"""

from __future__ import annotations

from typing import NamedTuple

from .db import Db
from .errors import NotAContainer, NotFound
from .types import NodeId, NodeType, Path

_WALK = "resolution.resolve_path"


class Resolved(NamedTuple):
    node: NodeId
    type: NodeType


def split_path(path: Path | str) -> list[str]:
    """Normalize a mount-relative path into clean segments.

    '' and '/' -> []. Collapses repeated/trailing slashes. No '.'/'..' handling:
    they are treated as ordinary names and will simply NotFound. That is not an
    omission to fix later — it is what keeps resolution anchored at the mount
    point (see the module docstring).
    """
    return [seg for seg in str(path).split("/") if seg]


def _walk(db: Db, root: NodeId, segments: list[str]) -> Resolved:
    """Resolve non-empty `segments` beneath `root` in one query.

    Raises the same errors, at the same segment, as the per-segment fold did:
    the CTE records where it stopped so the last row carries the diagnosis.
    """
    rows = db.all(_WALK, {"root": root, "path": "/".join(segments)})
    last = rows[-1]  # the walk always consumes at least one segment
    if last["node_id"] is None:
        raise NotFound(
            f"no visible child {last['seg']!r}",
            container=NodeId(last["parent_id"]),
            name=last["seg"],
        )
    if last["idx"] < len(segments):
        # stopped early: only a non-container can halt the walk mid-path
        raise NotAContainer(
            f"path segment {last['seg']!r} is not a container",
            node=NodeId(last["node_id"]),
        )
    return Resolved(NodeId(last["node_id"]), NodeType(last["node_type"]))


def resolve(db: Db, mount_point: NodeId, path: Path | str) -> Resolved:
    """Resolve a full mount-relative path to its node.

    `mount_point` is the resolved anchor for this mount (caller supplies it,
    typically MountInfo.mount_point). Returns the node and its type so callers
    can enforce container/entry expectations without a second lookup.
    """
    segments = split_path(path)
    if not segments:
        # '' or '/' is the mount point itself; report its type for symmetry.
        row = db.one("resolution.get_node", {"node": mount_point})
        if row is None:
            raise NotFound("mount point does not exist", node=mount_point)
        return Resolved(NodeId(mount_point), NodeType(row["type"]))
    return _walk(db, mount_point, segments)


class Parent(NamedTuple):
    container: NodeId
    name: str


def resolve_parent(db: Db, mount_point: NodeId, path: Path | str) -> Parent:
    """Resolve a path to (its parent container, its final name).

    The substrate for create_container / create_entry / move-target / rename.
    Resolving the parent walks all-but-last segments and requires the parent to
    be a container; the final name is NOT looked up (it may not exist yet).
    """
    segments = split_path(path)
    if not segments:
        raise NotFound("cannot take the parent of the mount point root")

    *head, final = segments
    if not head:
        return Parent(mount_point, final)  # parent is the mount point itself

    found = _walk(db, mount_point, head)
    if found.type is not NodeType.CONTAINER:
        # the deepest head segment resolved to an entry; _walk only raises
        # NotAContainer for segments it had to descend THROUGH
        raise NotAContainer(
            f"path segment {head[-1]!r} is not a container", node=found.node
        )
    return Parent(found.node, final)


__all__ = ["Resolved", "Parent", "split_path", "resolve", "resolve_parent"]
# Copyright Michael Godfrey 2026 | aloecraft.org <michael@aloecraft.org>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

File: rust/aloelite-core/src/resolve.rs
//! Path resolution — the first thing every implementation writes, and the
//! thing the whole flat layer is built on.
//!
//! [`resolve`] walks a whole path in ONE query via the
//! `resolution.resolve_path` recursive CTE, starting at the mount's mount
//! point. [`resolve_parent`] is the same walk stopped one segment short,
//! returning (container, final name) for the create / move / rename
//! operations that need "the parent container, keep the final name".
//!
//! Path semantics are decided HERE, once, so every implementation mirrors
//! exactly one set of rules:
//!   * paths are mount-relative; `''` and `'/'` both denote the mount point
//!   * leading/trailing slashes are ignored; empty segments (`//`) collapse
//!   * resolution sees only VISIBLE nodes (NODE-5) — the greatest-uuid7
//!     sibling wins — so hidden same-name siblings are unreachable by path;
//!     `*_by_id` variants exist for those
//!   * a miss at any segment is `not_found`
//!   * a non-final segment that resolves to a leaf is `not_a_container`
//!   * `.` and `..` are ORDINARY NAMES, not navigation. They are never
//!     interpreted, so a path can never climb above the mount point it
//!     started at — which is what confines a subtree mount to its subtree.
//!     That containment is a security boundary.

use rusqlite::named_params;

use crate::db::Db;
use crate::errors::{FsError, Result};
use crate::templates::resolution::{GET_NODE, RESOLVE_PATH};
use crate::types::{NodeId, NodeType};

// ---------------------------------------------------------------------------
// surface
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Resolved {
    pub node: NodeId,
    pub kind: NodeType,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Parent {
    pub container: NodeId,
    pub name: String,
}

/// Normalize a mount-relative path into clean segments. `''` and `'/'` give
/// an empty list; repeated and trailing slashes collapse. No `.`/`..`
/// handling: they are ordinary names and will simply not be found.
pub fn split_path(path: &str) -> Vec<&str> {
    path.split('/').filter(|s| !s.is_empty()).collect()
}

/// Resolve a full mount-relative path to its node and type, so callers can
/// enforce container/leaf expectations without a second lookup.
pub fn resolve(db: &mut Db, mount_point: &NodeId, path: &str) -> Result<Resolved> {
    let segments = split_path(path);
    if segments.is_empty() {
        // '' or '/' is the mount point itself; report its type for symmetry.
        let kind = db
            .one(GET_NODE, named_params! { ":node": mount_point }, |r| {
                Ok(r.get::<_, NodeType>("type")?)
            })?
            .ok_or_else(|| {
                FsError::not_found(format!("mount point {mount_point} does not exist"))
            })?;
        return Ok(Resolved {
            node: mount_point.clone(),
            kind,
        });
    }
    walk(db, mount_point, &segments)
}

/// Resolve a path to (its parent container, its final name). The parent is
/// walked and must be a container; the final name is NOT looked up (it may
/// not exist yet).
pub fn resolve_parent(db: &mut Db, mount_point: &NodeId, path: &str) -> Result<Parent> {
    let segments = split_path(path);
    let Some((final_name, head)) = segments.split_last() else {
        return Err(FsError::not_found(
            "cannot take the parent of the mount point root",
        ));
    };
    if head.is_empty() {
        // the parent is the mount point itself
        return Ok(Parent {
            container: mount_point.clone(),
            name: (*final_name).to_owned(),
        });
    }
    let found = walk(db, mount_point, head)?;
    if found.kind != NodeType::Container {
        // the deepest head segment resolved to a leaf; walk() only reports
        // not_a_container for segments it had to descend THROUGH
        return Err(FsError::NotAContainer { node: found.node.0 });
    }
    Ok(Parent {
        container: found.node,
        name: (*final_name).to_owned(),
    })
}

// ---------------------------------------------------------------------------
// depth: the one-query walk
// ---------------------------------------------------------------------------

struct WalkRow {
    idx: i64,
    seg: Option<String>,
    parent_id: Option<String>,
    node_id: Option<NodeId>,
    node_type: Option<String>,
}

/// Resolve non-empty `segments` beneath `root` in one query. The CTE records
/// where it stopped, so the last row carries the diagnosis.
fn walk(db: &mut Db, root: &NodeId, segments: &[&str]) -> Result<Resolved> {
    let rows = db.all(
        RESOLVE_PATH,
        named_params! { ":root": root, ":path": segments.join("/") },
        |r| {
            Ok(WalkRow {
                idx: r.get("idx")?,
                seg: r.get("seg")?,
                parent_id: r.get("parent_id")?,
                node_id: r.get("node_id")?,
                node_type: r.get("node_type")?,
            })
        },
    )?;
    // the walk always consumes at least one segment
    let last = rows
        .last()
        .ok_or_else(|| FsError::internal("resolve_path returned no rows"))?;
    let Some(node) = &last.node_id else {
        return Err(FsError::not_found(format!(
            "no visible child {:?} in {}",
            last.seg.as_deref().unwrap_or(""),
            last.parent_id.as_deref().unwrap_or("?")
        )));
    };
    if (last.idx as usize) < segments.len() {
        // stopped early: only a non-container can halt the walk mid-path
        return Err(FsError::NotAContainer {
            node: node.0.clone(),
        });
    }
    let kind = last
        .node_type
        .as_deref()
        .and_then(NodeType::parse)
        .ok_or_else(|| FsError::corrupt(format!("node {node} has an unknown type")))?;
    Ok(Resolved {
        node: node.clone(),
        kind,
    })
}

#[cfg(test)]
mod tests {
    use super::split_path;

    #[test]
    fn split_collapses_slashes_and_keeps_dots_as_names() {
        assert_eq!(split_path(""), Vec::<&str>::new());
        assert_eq!(split_path("/"), Vec::<&str>::new());
        assert_eq!(split_path("//a///b/"), vec!["a", "b"]);
        assert_eq!(split_path("/a/../b"), vec!["a", "..", "b"]);
    }
}
